File: src/test/java/im/tox/gui/FileTransferOutgoing.java
package im.tox.gui;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.Arrays;

public final class FileTransferOutgoing extends FileTransfer {

  private RandomAccessFile input;

  public FileTransferOutgoing(File file) {
    super(file);
  }

  @Override
  public void resume() throws FileNotFoundException {
    this.input = new RandomAccessFile(file, "r");
  }

  @Override
  public byte[] read(long position, int length) throws IOException {
    input.seek(position);
    byte[] data = new byte[length];
    int readLength = input.read(data);
    if (data.length > readLength) {
      data = Arrays.copyOf(data, readLength);
    }
    return data;
  }

  @Override
  public void close() throws IOException {
    input.close();
  }

  @Override
  public void write(long position, byte[] data) throws IOException {
    throw new IOException("Cannot write to outgoing file");
  }

}

File: projects/tox4j/src/test/java/im/tox/gui/domain/FileTransferOutgoing.scala
package im.tox.gui.domain

import java.io.{ File, FileNotFoundException, IOException, RandomAccessFile }

final class FileTransferOutgoing(file: File) extends FileTransfer(file) {
  private var input: RandomAccessFile = null

  @throws[FileNotFoundException]
  def resume(): Unit = {
    input = new RandomAccessFile(file, "r")
  }

  @throws[IOException]
  def read(position: Long, length: Int): Array[Byte] = {
    input.seek(position)

    val data = Array.ofDim[Byte](length)
    val readLength = input.read(data)

    if (data.length > readLength) {
      data.slice(0, readLength)
    } else {
      data
    }
  }

  @throws[IOException]
  def close(): Unit = {
    input.close()
  }

  @throws[IOException]
  def write(position: Long, data: Array[Byte]): Unit = {
    throw new IOException("Cannot write to outgoing file")
  }
}
